Fix disconnect / initialization

galfert opened the pull request:
refs #885

This bugfix is still work in progress.

It fixes problems when running `remoteStorage.disconnect()`, which prevented proper initialization and reconnection afterwards without reloading the Browser.

For some of the fixes there are still tests missing.

Also, there is still some timing issue which sometimes prevents the proper starting of the sync after calling configure (as described in the referenced issue).

skddc commented:
I reviewed the code changes and tested a build with Webmarks. LGTM! To be honest, I don't know if there's much more we can unit-test there for now.

Do you know about cause of the timing issue already?

galfert commented:
I added some more tests and fixed the actual problem of remoteStorage not syncing after configuring/connecting the remote.

The problem was that sync only started automatically when the remote is already configured at the time the `Sync` instance gets created. This is always the case when either (re-)loading an app when already connected (because the data gets directly loaded from localStorage), or when using the OAuth dance via the widget (there the remote gets configured before `Sync` gets initialized after the redirect).

Now, `Sync` listens on the `connected` event and starts syncing then.

This PR is ready for review/test now.

skddc commented on src/authorize.js:
```diff
@@ -48,11 +48,6 @@
-          // TODO
-          // sync doesn't start until after reload
-          // possibly missing some initialization step?
-          global.location.reload();
```
![great-success](https://cloud.githubusercontent.com/assets/842/15085636/4675199a-1390-11e6-9432-f85b20006c5a.jpg)

skddc commented:
The code looked good to me at first, but when I tested it with Webmarks, I found that the connected and disconnected events no longer work correctly. Here's a video demontrating it:

https://storage.5apps.com/basti/public/shares/160506-2059-rsjs-test-1.webm

(Don't try to skip back and forth, we need to implement content-range requests in 5apps Storage ;))

galfert commented:
Thanks for testing. I'll try to find out what's wrong there.

galfert commented:
FYI: I couldn't reproduce this in Chrome, but I'm seeing the same behavior in Firefox. Currently trying to find out what's wrong here.

galfert commented:
Took a while to find the cause, but I think I got that sucker :)

Reason was that the list of event handlers was modified while the event emitter was iterating over it, and thus skipped over one entry.

I think this should all finally be good to go. Thanks again for pointing out the issue.

skddc commented:
Nice catch! Will test again asap.

skddc commented:
Works now and lgtm. Good job!

silverbucket commented:
Looks good, I reviewed the code and did some basic tests and didn't run into any problems. +1

galfert commented:
Thanks for testing and reviewing, guys.
